libunwind: make it a virtual package

goxberry opened the pull request:
This pull request attempts to implement the suggested steps outlined in #8823 to make `libunwind` a virtual package.

jcftang commented:
AFAIK the likes of gperftools depend on libunwind

tgamblin commented:
> AFAIK the likes of gperftools depend on libunwind

The LLVM one or just `libunwind`

goxberry commented:
@tgamblin: re: your 1st point: this PR renames the current concrete `libunwind` package to `nongnu-libunwind`, so after this PR, there should be no concrete package named `libunwind`, and thus no namespace collision. I renamed the concrete `libunwind` package to `nongnu-libunwind` to avoid modifying every`depends_on('libunwind')` statement in spack, which may or may not lead the concretizer to reinstall many packages on user systems. Given this explanation, if you still want me to rename the virtual package to `unwind`, I’m happy to do so.

Re: your 2nd point: the new `nongnu-libunwind` provides `libunwind@1`; `llvm+internal_unwind` and `apple-libunwind` provide `libunwind@0`. So your proposal would be to replace `libunwind@1` with `nongnu-libunwind`, and have all of the libraries provide `libunwind` (or `unwind`, depending on your preference), and that’s reasonable. I don’t know which packages require this specific version of libunwind yet. 

Re: your 3rd point: sounds good to me. Will make that change (and the others you suggest) later today/early tomorrow, once I get back in front of a computer.

v-dobrev commented:
I think this PR is a good solution for the `libunwind` issues as summarized in #8823.

For the `mfem` package, the only change should be to replace the dependency on `libunwind` with the new virtual dependency `unwind`. We can do that in a separate PR but it will be easier to just add it here. What's the Spack policy for cases like this - add the change here, or in a separate PR?

tgamblin commented:
I’d just add it here to be transactional.

v-dobrev commented on var/spack/repos/builtin/packages/apple-libunwind/package.py:
```diff
@@ -0,0 +1,77 @@
+
+    provides('libunwind@0')
+
+    conflicts('platform!=darwin')
```
This line did not work for me. I switched to
```
    depends_on('platform=darwin')
```
which worked fine on Mac and I expect will reject any other platform.

v-dobrev replied:
I added a `print` statement to see what `word` is (just before the `raise` command) and I got that word is `platform!=darwin`.

goxberry commented:
@tgamblin: An update: turns out LLVM doesn't install `libunwind.h`, only `unwind.h`. So I dropped the `provides('unwind')` directive in the spack `llvm` package. That said, LLVM still has a libunwind implementation that provides `libunwind.h` (https://github.com/llvm-mirror/libunwind), so a virtual package still seems like a reasonable design choice in the event that someone installs that header.

A read of the `conflicts` directive documentation indicates its argument must be a valid spec, so instead of negating `platform=darwin`, I listed all of the platforms in http://spack.readthedocs.io/en/latest/spack.platforms.html, which resolves the bug @v-dobrev encountered.

For now, I've updated `mfem`, `gperftools`, and `caliper`. The other packages that depend on `libunwind` currently can't be built on macOS. `cbtf-krell` and `extrae` both depend on `dyninst`, which I believe is incompatible with macOS. It should be possible to build `mpip` with macOS -- I've done it --  but the `libdwarf` package is hosed on macOS right now due to a bad linker command.

v-dobrev commented:
It looks like some of the travis checks fail because the virtual package `unwind` does not have a default provider in `etc/spack/defaults/packages.yaml`.

goxberry commented:
Thanks for the heads up @v-dobrev! These issues should be fixed with [c50fb69](https://github.com/spack/spack/pull/8825/commits/c50fb69549451887244ea1889164c2dada4c5049), which I just pushed.

tgamblin commented on etc/spack/defaults/darwin/packages.yaml:
```diff
@@ -18,3 +18,10 @@ packages:
+    # Apple bundles libunwind version 35.3 with macOS 10.9 and later,
+    # although the version number used here isn't critical
+      apple-libunwind@35.3: /usr
+    buildable: False
```
nice.

tgamblin commented on etc/spack/defaults/packages.yaml:
```diff
@@ -39,3 +39,4 @@ packages:
       scalapack: [netlib-scalapack]
       szip: [libszip, libaec]
       tbb: [intel-tbb]
+      unwind: [libunwind]
```
This might be the most minor change request I've ever put in on a PR, but it would be nice to have a newline here.

goxberry replied:
Added in [3099599](https://github.com/spack/spack/pull/8825/commits/3099599033a524cabfa56bed405a4a0fae072b86).
